docs/proposals/20200720-single-controller-multitenancy.md

```diff
@@ -0,0 +1,548 @@
+  the principal type.
+* The controller will compare the hash of the credential provider against the same secret’s provider
+  in a cache ([NFR2](#NFR2)).
+* The controller will take the newer of the two and instantiate Azure clients with the selected 
+  `Authorizer`.
+* The controller will set the Identity resource as one of the OwnerReferences of the AzureCluster.
+
+### Clusterctl changes
+
+Today, `clusterctl move` operates by tracking objectreferences within the same namespace, since we 
+are now proposing to use cluster-scoped resources, we will need to add requisite support to 
+clusterctl's object graph to track ownerReferences pointing at cluster-scoped resources, and ensure 
+they are moved. We will naively not delete cluster-scoped resources during a move, as they maybe 
+referenced across namespaces.
+
+#### Validating webhook changes
+
+None.
+
+A validating webhook could potentially handle some cross-resource validation necessary for the [security
+model](#security-model) and provide more immediate feedback to end users. However, it would be 
+imperfect. For example, a change to an Azure Identity resource could affect the validity of 
+corresponding AzureCluster.
+
+#### Principal Type Credential Provider Behaviour
+
+Implementations for all principal types will implement the AutoREST `Authorizer` interface as well 
+as an additional function signature to support caching.
+
+```go
+// Authorizer is the interface that provides a PrepareDecorator used to supply request
+// authorization. Most often, the Authorizer decorator runs last so it has access to the full
+// state of the formed HTTP request.
+type Authorizer interface {
+    WithAuthorization() PrepareDecorator
+}
+
+type AzureIdentityProvider interface {
+    Authorizer
+
+    // Hash returns a unique hash of the data forming the credentials
+    // for this principal
+    Hash() (string, error)
+}
+```
+
+Azure client sessions are structs implementing the `Authorizer` interface. Every SDK will verify the
+token in the `Authorizer` has not expired. If the token has expired, the `Authorizer` will attempt
+to fetch a new token from AAD.
+
+The controller will maintain a cache of all the `Authorizer`s used by the clusters.
+
+In the first instance, the controller will refresh secrets on each reconciliation run, and store a 
+hash of the secret contents in the cache. If the hash changes, then the `Authorizer` will be
+recreated via the secret contents.
+
+This could be further optimised by having the controller maintain a watch on all Secrets matching 
+the principal types. Upon receiving an update event, the controller will update lookup the key in 
+the cache and update the relevant `Authorizer`. This may be implemented as its own interface. 
+Mutexes will ensure in-flight updates are completed prior to SDK calls are made. This would require 
+changes to RBAC, and maintaining a watch on secrets of a specific type will require further
```

Comment:
> maintaining a watch on secrets of a specific type will require further
investigation as to feasibility

this sounds fine to me, anything you're worried about? if there's any identifying metadata you can use it to write a filter

Reply:
I was not, but the CAPA authors were. Not sure if I missed something... Wanted to leave it in for someone smarter than me to spot the hole.